substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java

```diff
@@ -0,0 +1,545 @@
+import com.oracle.svm.hosted.snippets.ReflectionPlugins;
+
+import jdk.vm.ci.meta.ResolvedJavaMethod;
+
+/**
+ * This plugin analyses the graph for the resolved Java method and specifies what should to be
+ * inlined during graph parsing before the static analysis. Plugin is searching for the methods that
+ * folds to a constant or an array of constants.
+ * <p>
+ * Usage: To use this plugin, add option -H:+InlineBeforeAnalysis to the native-image.
+ *
+ * <h2>Structure of the NativeImageInlineDuringParsingPlugin</h2>
+ *
+ * <ul>
+ *
+ * <li>InvocationResultInline -- stores inline data for the currently parsed invoke (resolved Java
+ * method).</li>
+ *
+ * <li>CallSite -- class that stores information about caller method. This information are used in
+ * InvocationResultInline and saved in the dataInline map from where we read information after
+ * analysis.</li>
+ *
+ * <li>TrivialMethodDetector -- class that analyses the callee graph and detects if it should be
+ * inlined or not.</li>
+ *
+ * <li>MethodNodeTrackingAndInline -- innerclass of the TrivialMethodDetector that collects
+ * information during graph construction to filter non-trivial methods and inline trivial invokes
+ * (callees children). It is searching for the method that folds to constant, so an unlimited amount
+ * of constants, parameter nodes and load filed nodes is allowed in the callees graph. On the other
+ * hand, we don't want to inline if the instance node, array node, store filed node is detected, and
+ * also we don't allow call target nodes and invokes that don't fold to constant. During graph
+ * analysis, this class also inlines trivial invokes (callees children).</li>
+ *
+ * </ul>
+ * <p>
+ * The results of an inlining decision are placed in the
+ * {@link NativeImageInlineDuringParsingSupport#inlineData}.
+ * <p>
+ * Example: Assume that we have a graph with methods R, A, B, C, where R is the root method, A calls
+ * B, B calls C, and C returns the constant 1.*
+ *
+ * <pre>
+ *       R
+ *      /
+ *     A
+ *    /
+ *   B
+ *  /
+ * C   <-- only returns 1
+ * </pre>
+ * <p>
+ * We first analyze A and decide whether to enter the method for further analysis or not. In first
+ * case, we proceed and analyze the nodes deeper in the graph with MethodNodeTrackingAndInline. We
+ * repeat the procedure with B and C, and when the analysis of method C is complete, we decide to
+ * inline it because it only returns a constant value. After that, B folds to constant, so we inline
+ * this method too. Finally, for the same reason, we decide to inline A into R.
+ */
+@SuppressWarnings("ThrowableNotThrown")
+public class NativeImageInlineDuringParsingPlugin implements InlineInvokePlugin {
+
+    public static class Options {
+        @Option(help = "Inline methods which folds to constant during parsing before the static analysis.")//
+        public static final HostedOptionKey<Boolean> InlineBeforeAnalysis = new HostedOptionKey<>(true);
```

Comment:
If we do not enable the option by default in this PR, then the option help text should start with "Experimental: "

Reply:
Done.